File: scripts/eaglercraft.py
js.document.body.innerHTML = ''
js.document.title = 'Eaglercraft 1.13'

sounds = dict(literal_eval(sounds))

DEBUGGING = True
if 'debugging=true' in dev_file:
    DEBUGGING = True


def broadcast(text):
    if DEBUGGING:
        now = datetime.datetime.now()
        print(f'[EAGLERCRAFT] ({now}): {text}')

broadcast('Eaglercraft 1.13 client started.')

eagler_words = ['Eagler', 'Darv', 'Yigg', 'Deev', 'Vigg', 'Yeer', 'Yeeish']
USERNAME = random.choice(eagler_words) + random.choice(eagler_words) + str(random.randint(20, 9999))
broadcast(f'Username: {USERNAME}')

eagwrite = EagWrite()

windowW = int(windowW)
windowH = int(windowH)

is_agreement_loaded = False

js.window.addEventListener('error', lambda event: js.console.error(f"Global Error: {event.message}"))

class AudioManager:
    def __init__(self):
        self.playing_audio = []

    def play(self, sound_data_base64):
        sound_url = f"data:audio/ogg;base64,{sound_data_base64}"
        try:
            self.sound = js.Audio.new(sound_url)
            self.sound.loop = False
            self.sound.play()
        except js.JsException as e:
            js.console.error(f"Error playing sound: {e}")
        self.playing_audio.append(self.sound)
        return self.sound

    def stop(self, audio_data):
        if audio_data is not None:
            audio_data.pause()
            audio_data.currentTime = 0
        self.playing_audio.remove(self.sound)

    def toggle_loop(self, audio_data):
        if audio_data is not None:
            audio_data.loop = True

    def adjust_volume(self, volume, audio_data):
        if audio_data is not None:
            audio_data.volume = volume

    def stop_all_sound(self):
        for sound in self.playing_audio:
            self.stop(sound)

audio = AudioManager()

audio.play(sounds['menu1'])

is_agreement_loaded = False

event_proxies = []
start_buttons = []
start_text = []

def create_button(text='Button', size=2, onclick=None, x=100, y=100, text_tune=70):
    y += 20
    text_x = x + 50  # Adjust for general centering of text inside button
    text_y = y - 5

    js.console.log(f"Creating button with text '{text}' at position ({x}, {y})")

    def click(event):
        js.console.log("Button clicked!")
        if hasattr(audio, 'playing') and audio.playing:
            audio.stop()
        audio.play(sounds['click'])
        if onclick is not None:
            onclick(event)

    def hover(event):
        nonlocal button_text
        eagwrite.change_color(button_text, '#feffa0')

    def leave(event):
        nonlocal button_text
        eagwrite.change_color(button_text, 'white')

    click_proxy = create_once_callable(click)

    hover_proxy = create_proxy(hover)
    leave_proxy = create_proxy(leave)
    event_proxies.extend([click_proxy, hover_proxy, leave_proxy])

    button = js.document.createElement('button')
    button.className = 'gui-widgets-button'
    button.style.transform = f'scale({size})'
    button.style.position = 'absolute'
    button.style.left = f'{x}px'
    button.style.top = f'{y}px'
    button.addEventListener('click', click_proxy)
    button.addEventListener('mouseenter', hover_proxy)
    button.addEventListener('mouseleave', leave_proxy)
    js.document.body.appendChild(button)

    # Adjust text centering inside the button
    button_text = eagwrite.write(text, x + text_tune, text_y, 'white', size=size + 0.2)

    def destroy():
        if button and button.parentNode:
            js.console.log("Removing button element from DOM.")
            js.document.body.removeChild(button)
        else:
            js.console.warn("Button element not found in DOM during removal.")
        
        eagwrite.destroy(button_text)
        
        click_proxy.destroy()
        hover_proxy.destroy()
        leave_proxy.destroy()

        event_proxies.remove(click_proxy)
        event_proxies.remove(hover_proxy)
        event_proxies.remove(leave_proxy)

    return [button_text, button, destroy]

def destroy_button(button_data):
    try:
        js.console.log("Destroying button.")
        eagwrite.destroy(button_data[0])
        
        if button_data[1].parentNode:
            button_data[1].remove()
        else:
            js.console.warn("Button element not found in DOM during removal.")

        button_data[2]()
    except Exception as e:
        js.console.error(f"Error destroying button: {e}")

def load_agreement():
    global is_agreement_loaded
    if is_agreement_loaded:
        return
    is_agreement_loaded = True
    js.console.log("Loading agreement screen.")
    js.document.body.style.backgroundImage = "url('assets/eagler/editprofile_bg.png')"
    text1 = eagwrite.write('&l Information', windowW / 2 - 140, 200, 'lime', shadow_color='#383838', size=5) 
    text2 = eagwrite.write('&o This software is NOT pirated', windowW / 2 - 300, 300, 'red', shadow_color='#383838', size=5) 
    text3 = eagwrite.write('&n This is just a RECREATION of Minecraft', windowW / 2 - 400, 400, 'yellow', shadow_color='#383838', size=5) 
    text4 = eagwrite.write('Any DMCA request toward this client is unnecessary', windowW / 2 - 700, 500, 'orange', shadow_color='#383838', size=5)

    def agreement_profile_page(_):
        js.console.log("Moving to edit profile page.")
        eagwrite.destroy(text1)
        eagwrite.destroy(text2)
        eagwrite.destroy(text3)
        eagwrite.destroy(text4)
        destroy_button(agreement_button)
        edit_profile_page()

    agreement_button = create_button('Next', x=windowW / 2 - 100, y=windowH - 300, onclick=agreement_profile_page)
    start_buttons.append(agreement_button)

def edit_profile_page(_=None):
    js.document.body.style.backgroundImage = "url('assets/eagler/editprofile_bg.png')"
    edit_profile_text = eagwrite.write(text='Edit Profile', x=windowW / 2 - 140, y=0, color='black', size=5, shadow_color='#383838')

    start_text.append(edit_profile_text)

    done_button = create_button('Done', x=windowW / 2 - 100, y=windowH - 200, onclick=titlescreen.open_titlescreen)
    
    import_skin_button = create_button('Import Skin', x=windowW / 2 - 100, y=windowH - 300, text_tune=40)

    start_buttons.append(done_button)
    start_buttons.append(import_skin_button)

    editprofile_frame = js.document.createElement('div')
    editprofile_frame.style.position = 'absolute'
    editprofile_frame.style.backgroundColor = 'rgb(0, 0, 21)'
    editprofile_frame.style.border = '2px solid yellow'
    editprofile_frame.style.width = '250px'
    editprofile_frame.style.height = '400px'
    editprofile_frame.style.top = '200px'
    editprofile_frame.style.left = f'{windowW/2-150}px'
    js.document.body.appendChild(editprofile_frame)

    dropdown = js.document.createElement('button')
    dropdown.style.position = 'absolute'
    dropdown.className = 'eagler-dropdown'
    dropdown.style.top = '350px'
    dropdown.style.left = f'{windowW/2+150}px'
    js.document.body.appendChild(dropdown)

load_agreement()

class Titlescreen:
    def __init__(self):
        pass

    def clear_start(self):
        for a in start_buttons:
            destroy_button(a)
        start_buttons.clear()

        for b in start_text:
            eagwrite.destroy(b)
        start_text.clear()

    def open_titlescreen(self, _):
        self.clear_start()

global titlescreen
titlescreen = Titlescreen()

File: scripts/eaglercraft.js
function broadcast(text) {
    let now = new Date();
    console.log(`[EAGLERCRAFT] (${now}): ${text}`);
}

broadcast('Eaglercraft 1.20 client started.');

let chosen_skin = 'Default Steve';
let eagler_words = ['Eagler', 'Darv', 'Yigg', 'Deev', 'Vigg', 'Yeer'];
let USERNAME = eagler_words[Math.floor(Math.random() * eagler_words.length)] + 
               eagler_words[Math.floor(Math.random() * eagler_words.length)] + 
               (Math.floor(Math.random() * 79) + 20);
broadcast(`Username: ${USERNAME}`);

const panoramaImages = [
    'assets/gui/title/background/panorama_0.png',
    'assets/gui/title/background/panorama_1.png',
    'assets/gui/title/background/panorama_2.png',
    'assets/gui/title/background/panorama_3.png',
    'assets/gui/title/background/panorama_4.png',
    'assets/gui/title/background/panorama_5.png'
];

const splashTexts = [
    'eaglercraft 1.20',
    'woohooo!',
    'You Eagler!',
    'You Vigg!',
    'You Darviggler!',
    'You Raven!',
    'Play it Now!',
]

function getRandomItem(arr) {
    const randomIndex = Math.floor(Math.random() * arr.length);
    return arr[randomIndex];
}

function launchEaglercraft() {
    document.body.innerHTML = '';
    

    let windowW = window.innerWidth;
    let windowH = window.innerHeight;


    let back = 0;
    if (windowW < 1920) {
       back = windowH - 250
    }

    console.log(windowW, windowH)

    let is_agreement_loaded = false;

    window.addEventListener('error', (event) => {
        console.error(`Global Error: ${event.message}`);
    });

    class AudioManager {
        constructor() {
            this.playing_audio = [];
        }

        play(sound_url) {
            try {
                const sound = new Audio(sound_url);
                sound.loop = false;
                sound.play();
                this.playing_audio.push(sound);
                return sound;
            } catch (e) {
                console.error(`Error playing sound: ${e}`);
            }
        }

        stop(audio_data) {
            if (audio_data) {
                audio_data.pause();
                audio_data.currentTime = 0;
                this.playing_audio = this.playing_audio.filter(sound => sound !== audio_data);
            }
        }

        toggle_loop(audio_data) {
            if (audio_data) {
                audio_data.loop = !audio_data.loop;
            }
        }

        adjust_volume(volume, audio_data) {
            if (audio_data) {
                audio_data.volume = Math.min(Math.max(volume, 0), 1); 
            }
        }

        stop_all_sound() {
            this.playing_audio.forEach(sound => this.stop(sound));
        }
    }

    let audio = new AudioManager();
    audio.play('assets/sounds/music/menu/menu1.ogg');

    let event_proxies = [];

    function create_button(text = 'Button', size = 2, onclick = null, x = 100, y = 100, text_tune = 70, state='gui-widgets-button', animation='none') {
        let button_text; 
        y += 20;
        let text_x = x + 50;
        let text_y = y - 5;
    
        console.log(`Creating button with text '${text}' at position (${x}, ${y})`);
    
        let click_proxy = (event) => {
            console.log("Button clicked!");
            audio.play('assets/sounds/random/click.ogg');
            if (onclick) onclick(event);
        };
    
        let hover_proxy = () => {
            eagwrite.change_color(button_text, '#feffa0');
        };
    
        let leave_proxy = () => {
            eagwrite.change_color(button_text, 'white');
        };
    
        let button = document.createElement('button');
        button.className = state;
        button.style.transform = `scale(${size})`;
        button.style.position = 'absolute';
        button.style.animation = animation;
        button.style.left = `${x}px`;
        button.style.top = `${y}px`;
        button.addEventListener('click', click_proxy);
        button.addEventListener('mouseenter', hover_proxy);
        button.addEventListener('mouseleave', leave_proxy);
        document.body.appendChild(button);
    
        if (state == 'gui-widgets-button') {
            button_text = eagwrite.write(text, x + text_tune, text_y, 'white', '#383838', size + 0.5, 3, 1, animation); 
        } else {
            button_text = eagwrite.write(text, x + text_tune, text_y, 'rgb(121, 121, 121)', null, size + 0.5, 3, 1, animation);
        }
    
        function destroy() {
            if (button && button.parentNode) {
                console.log("Removing button element from DOM.");
                document.body.removeChild(button);
            } else {
                console.warn("Button element not found in DOM during removal.");
            }
            eagwrite.destroy(button_text);
        }
    
        return [button_text, button, destroy];
    }

    function destroy_button(button_data) {
        try {
            console.log("Destroying button.");
            eagwrite.destroy(button_data[0]);
            button_data[1].remove();
            button_data[2]();
        } catch (e) {
            console.error(`Error destroying button: ${e}`);
        }
    }

    function load_agreement() {
        if (is_agreement_loaded) return;
        is_agreement_loaded = true;
        console.log("Loading agreement screen.");
        document.body.style.backgroundImage = `url('assets/eagler/bg/${Math.floor(Math.random() * 7) + 1}.png')`;
        let text3 = eagwrite.write('&n This is just a RECREATION of Minecraft', windowW / 2 - 400, 400, 'yellow', '#383838', 5);
        let text4 = eagwrite.write('Any DMCA request toward this client is unnecessary', windowW / 2 - 700, 500, 'orange', '#383838', 5);


        function agreement_profile_page() {
            console.log("Moving to edit profile page.");
            eagwrite.destroy(text1);
            eagwrite.destroy(text2);
            eagwrite.destroy(text3);
            eagwrite.destroy(text4);
            destroy_button(agreement_button);
            edit_profile_page();
        }

        let agreement_button = create_button('Agree', 2, titlescreen.open_titlescreen.bind(titlescreen), windowW / 2 - 100, windowH - 300);
    }
        input.focus();
    }

    class Titlescreen {
        clear_start(_) {
            console.log(USERNAME)
            document.body.innerHTML = ''
        }

        open_titlescreen() {
            let splashText = getRandomItem(splashTexts)
            this.clear_start();
            this.panorama();

            const minecraft_photo = document.createElement('img');
            minecraft_photo.src = 'assets/gui/title/minecraft.png';
            minecraft_photo.style.position = 'absolute';
            minecraft_photo.style.width = '500px';
            minecraft_photo.style.left = '50%';
            minecraft_photo.style.top = '100px';
            minecraft_photo.style.animation = 'fadeIn 2s forwards';
            minecraft_photo.style.transform = 'translate(-50%, -50%)'; 
            document.body.appendChild(minecraft_photo); 

            const java_photo = document.createElement('img');
            java_photo.src = 'assets/gui/title/edition.png';
            java_photo.style.position = 'absolute';
            java_photo.style.width = '250px';
            java_photo.style.left = '50%';
            java_photo.style.top = '140px';
            java_photo.style.animation = 'fadeIn 2s forwards';
            java_photo.style.transform = 'translate(-40%, -60%)';
            document.body.append(java_photo)

            eagwrite.expandContract(eagwrite.write(splashText, x=windowW/2, y=170, 'rgb(255, 255, 84)', '#383838', 3, 3, 1, 'none', -30), 1500, 'ease-in-out')

        
            eagwrite.write('Minecraft 1.20.1', 0, windowH-50, 'white', '#383838', 3, 3, 1, 'fadeIn 2s forwards');
            eagwrite.write('::RECREATION::', windowW-600, windowH-50, 'white', '#383838', 3);
            create_button('Singleplayer', 2, null, windowW/2-100, windowH/2-100, 20, 'gui-widgets-button', 'fadeIn 2s forwards');
            create_button('Multiplayer', 2, null, windowW/2-100, windowH/2-50, 30, 'gui-widgets-button', 'fadeIn 2s forwards');
            create_button('Mods', 2, null, windowW/2-100, windowH/2-50, 40, 'gui-widgets-button-disabled', 'fadeIn 2s forwards');
        }

        panorama() {
            const scene = new THREE.Scene();
            const camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000);
            const renderer = new THREE.WebGLRenderer({ alpha: true });
            renderer.setSize(window.innerWidth, window.innerHeight);
            document.body.appendChild(renderer.domElement);

            document.body.style.margin = '0';
            document.body.style.overflow = 'hidden';

            const textureLoader = new THREE.TextureLoader();

            const materials = [
                new THREE.MeshBasicMaterial({
                    map: textureLoader.load(panoramaImages[2]),
                    side: THREE.BackSide
                }),
                new THREE.MeshBasicMaterial({
                    map: textureLoader.load(panoramaImages[0]),
                    side: THREE.BackSide
                }),
                new THREE.MeshBasicMaterial({
                    map: textureLoader.load(panoramaImages[4]),
                    side: THREE.BackSide
                }),
                new THREE.MeshBasicMaterial({
                    map: textureLoader.load(panoramaImages[5]),
                    side: THREE.BackSide
                }),
                new THREE.MeshBasicMaterial({
                    map: textureLoader.load(panoramaImages[1]),
                    side: THREE.BackSide
                }),
                new THREE.MeshBasicMaterial({
                    map: textureLoader.load(panoramaImages[3]),
                    side: THREE.BackSide
                })
            ];

            const sphereGeometry = new THREE.BoxGeometry(500, 500, 500);
            const sphere = new THREE.Mesh(sphereGeometry, materials);
            scene.add(sphere);

            camera.position.set(0, 0, 0.1);

            const rotationSpeed = 0.001;

            function animate() {
                requestAnimationFrame(animate);
                
                sphere.rotation.y += rotationSpeed;  

                renderer.render(scene, camera);
            }

            animate();

            window.addEventListener('resize', () => {
                camera.aspect = window.innerWidth / window.innerHeight;
                camera.updateProjectionMatrix();
                renderer.setSize(window.innerWidth, window.innerHeight);
            });
        }
    }

    let titlescreen = new Titlescreen();

    load_agreement();
}

    
